src/internal/packager2/layout/package.go

```diff
@@ -0,0 +1,306 @@
+	"github.com/google/go-containerregistry/pkg/v1/layout"
+	"github.com/mholt/archiver/v3"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/sigstore/cosign/v2/cmd/cosign/cli/options"
+	"github.com/sigstore/cosign/v2/cmd/cosign/cli/verify"
+
+	"github.com/zarf-dev/zarf/src/api/v1alpha1"
+	"github.com/zarf-dev/zarf/src/config"
+	"github.com/zarf-dev/zarf/src/pkg/packager/deprecated"
+	"github.com/zarf-dev/zarf/src/pkg/transform"
+	"github.com/zarf-dev/zarf/src/pkg/utils"
+)
+
+// PackageLayout manages the layout for a package.
+type PackageLayout struct {
+	dirPath string
+	Pkg     v1alpha1.ZarfPackage
+}
+
+// PackageLayoutOptions are the options used when loading a package.
+type PackageLayoutOptions struct {
+	PublicKeyPath           string
+	SkipSignatureValidation bool
+	IsPartial               bool
+}
+
+// LoadFromTar unpacks the give compressed package and loads it.
+func LoadFromTar(ctx context.Context, tarPath string, opt PackageLayoutOptions) (*PackageLayout, error) {
+	dirPath, err := utils.MakeTempDir(config.CommonOptions.TempDirectory)
+	if err != nil {
+		return nil, err
+	}
+	err = archiver.Walk(tarPath, func(f archiver.File) error {
+		if f.IsDir() {
+			return nil
+		}
+		header, ok := f.Header.(*tar.Header)
+		if !ok {
+			return fmt.Errorf("expected header to be *tar.Header but was %T", f.Header)
+		}
+		// If path has nested directories we want to create them.
+		dir := filepath.Dir(header.Name)
+		if dir != "." {
+			err := os.MkdirAll(filepath.Join(dirPath, dir), helpers.ReadExecuteAllWriteUser)
+			if err != nil {
+				return err
+			}
+		}
+		dst, err := os.Create(filepath.Join(dirPath, header.Name))
+		if err != nil {
+			return err
+		}
+		defer dst.Close()
+		_, err = io.Copy(dst, f)
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	p, err := LoadFromDir(ctx, dirPath, opt)
+	if err != nil {
+		return nil, err
+	}
+	return p, nil
+}
+
+// LoadFromDir loads and validates a package from the given directory path.
+func LoadFromDir(ctx context.Context, dirPath string, opt PackageLayoutOptions) (*PackageLayout, error) {
+	b, err := os.ReadFile(filepath.Join(dirPath, ZarfYAML))
+	if err != nil {
+		return nil, err
+	}
+	var pkg v1alpha1.ZarfPackage
+	err = goyaml.Unmarshal(b, &pkg)
+	if err != nil {
+		return nil, err
+	}
+	if len(pkg.Build.Migrations) > 0 {
+		for idx, component := range pkg.Components {
+			pkg.Components[idx], _ = deprecated.MigrateComponent(pkg.Build, component)
+		}
+	}
+
+	pkgLayout := &PackageLayout{
+		dirPath: dirPath,
+		Pkg:     pkg,
+	}
+	err = validatePackageIntegrity(pkgLayout, opt.IsPartial)
+	if err != nil {
+		return nil, err
+	}
+	err = validatePackageSignature(ctx, pkgLayout, opt.PublicKeyPath, opt.SkipSignatureValidation)
+	if err != nil {
+		return nil, err
+	}
+	return pkgLayout, nil
+}
+
+// Cleanup removes any temporary directories created.
+func (p *PackageLayout) Cleanup() error {
```

Comment:
Does this increase readability?

Reply:
Compared to having callers call os.RemoveAll? I think it does, plus it enables us to do more clean up in the future if we need to.